### ISSUE TITLE
API: Add Section for Configuration and Querying parameters after establishment

### ISSUE THREAD

**Issue by theri** [taps-api/drafts#29]
Section 5.2 currently includes Connection.QueryProperties(), but it belongs to the Pre-Establishment Phase. At this point, most Protocol Properties are not available to be queried yet. So the current version of 5.2.2 only includes "Maximum Message Size Before Connection Establishment", as this clearly belongs in the Pre-Establishment phase.
There should probably be a "Configuration" section or similar that specifies what Protocol Properties can be queried (and set?) after establishment. Maybe the Protocol Properties should even be moved there altogether, but I like them in the Pre-Establishment section, because this makes it clear that applications should specify them as early as possible.

**Comment by tfpauly** [comment]
A few thoughts on the architecture implications of the properties here:

- Any properties that determine protocol or path selection, or need to be known at protocol handshake time, MUST be part of Pre-Establishment
- The set of protocol properties that may be set after establishment is a smaller set of knobs, and includes the per-send/receive options
- Querying parameters can reflect what was originally set on the connection; which is separate from what options are currently active. Getting both is useful.

The way I've built this currently is:

1. During Pre-Establishment, there is a large parameters object, which includes various protocol selection properties and protocol-specific options. Each protocol can define extensions to their own protocol options. At any point, the application can retrieve this original parameters object.

2. There is a separate protocol metadata object that defines the post-establishment state and knobs. This protocol metadata may be attached to a Send(), retrieved from a Receive(), or queried directly on a connection object outside of the data path once the connection is established. Any properties of the protocol that are eligible to be modified post-connection have knobs here. It is possible that some knobs show up in both pre-establishment and post-establishment, but one is explicitly intended to drive protocol and path selection, while the other is more about dynamic behavior.

**Comment by mwelzl** [comment]
So whatever the application configures later on a socket is only advisory anyway - too late to make strict requests. Hence we could perhaps simply say, in this document, that ALL protocol parameters are available for later configuration of a Connection (and implementations may choose to limit this to a subset). What do you think?

**Comment by tfpauly** [comment]
I really don't like the idea that all properties are available to change in the API after establishment, with the assumption that most of them do nothing or have undefined behavior.

If I say pre-establishment, "I prohibit cellular interfaces", and then later the API lets me express, "I require cellular interfaces"... what happens? You'll just ignore my requirement? Or the connection will suddenly fail?

If I say pre-establishment, "I allow unreliable transports", and then later the API lets me express, "I require reliable transports"... what happens? We fail the connection if we chose UDP?

I don't see any value in letting most properties be reset.

**Comment by mwelzl** [comment]
okay, that was just an idea for a possible easy way out  :-)

**Comment by theri** [comment]
Agree with Tommy - it makes no sense to allow these kinds of severe configuration errors in the API.
As protocol and path selection happen with connection establishment, I think both Protocol Selection Properties and Path Selection Properties should not be possible to change afterwards, only queried what was originally specified. On the other hand, Protocol Properties (not Selection, but configuration of the connection) and Intents can also be set later, either on the connection or on the individual Content. 
Protocol Properties are now spread across Section 5.2.2 (applied to Connections and Connection Groups) and 7.1 (applied to Content). I think this is fine. So maybe we do not need a Configuration section of its own in this draft after all.
In Issue #28 we came up with the idea of putting some information about properties, e.g., in which phase should they be set and whether they can be marked as "required", into the appendix. I think we could add content from this issue, like "Protocol Selection properties can only be queried after Create(), while Protocol Properties can be set" to that same appendix.

**Comment by theri** [comment]
Talking about this with Philipp, we found that it might make sense to allow setting a Path Selection Property even after connection establishment in case a multi-path protocol is being used. Not in the sense of "first I prohibit cellular, then I require cellular", which would still be a configuration error. But we see a case where the application first does not specify any path selection properties, an MPTCP connection is established with one subflow over cellular, one over another interface. Then it might make sense for the application to specify per-Content that some Content should not go over the cellular.
As I see it, the alternative would be to model the MPTCP connection as a connection group and the subflows as connections, so the application can send the "prohibit cellular" content over the non-cellular connection, which we chose not to do in #31, or to not have this option at all.

**Comment by tfpauly** [comment]
The point you make about MPTCP is potentially a use case (generally, I want to specify which policies I want for interface usage by MPTCP up front), but I think it only emphasizes the point I made before that we must separate out very specifically the properties set during pre-establishment from those after.

There is a large set of properties to be set at pre-establishment, around both path and protocol selection, and general protocol properties. In most cases, most of these properties are not eligible to be changed later. Thus exposing them in the same form later in a generic way will only lead to API misuse.

For the case of MPTCP, the fact that it may or may not be able to send certain data over certain sub flows going over interfaces is an advanced and optional feature of a specific transport protocol. This should be expressed, as I listed before, as per-content protocol metadata to influence the transport protocol that was selected. This may look like it overlaps with the properties set during pre-establishment, but from the application's view, it is now setting a specific MPTCP-send-property, rather than a connection-establishment-property. These are two very different namespaces, and we must leave them separate.

If we are not rigorous about being clear where properties are applicable, we are liable to design an API that either is non-deterministic based on which transports are selected or is full of knobs that must have detailed documentation about effects in different use cases. Better to have a fool-proof way of specifying things that focuses on making the general path as simple and transport-independent as possible.

**Comment by theri** [comment]
I agree, and I wrote a corresponding paragraph in Appendix A.2 of the current API draft.

So if we are not going to have a separate Configuration/Querying/Maintenance section, we can close this issue, right?

**Comment by britram** [comment]
IMO the MPTCP example illustrates that this isn't the right way to split up connection properties post-establishment. Let me suggest a different way to split this up (which might feed back into the architecture document, not sure yet):

- A Connection is configured with a set of properties, each bound to a preference ("require", "prohibit", "prefer", "avoid") before establishment. Until establishment, this connection is essentially *just* a container for its endpoint specifier(s) and properties. 

- Once a Connection is established, it is bound to one or more Protocol Stacks, some of which may disappear after racing is complete. Each of these protocol stacks also has properties, expressed in the same vocabulary, but without the preference bindings. For logging and cloning purposes, the Connection still maintains a read-only reference to its pre-establishment properties. 

- Each Protocol Stack also has a set of Properties, which correspond to the properties it can provide and/or has been configured with. These are defined in terms of the same vocabulary provided for pre-establishment, but are not associated with preferences. Some of these properties may be writable after (indeed, the post-establishment writability of a given property may itself be a transport property)

- Each Protocol Stack is bound to one or more Paths. Paths may also have properties,  derived from the local provisioning domain, online measurement done by the protocol stack, or other sources. Path properties are mainly made accessible to the application for logging purposes.

The key idea is that pre-establishment and post-establishment properties are fundamentally differently organized, though expressed using the same vocabulary, because they refer to fundamentally different views of similar (but not the same) concepts.

**Comment by britram** [comment]
I do think we at least want a separate "querying" section -- some applications are going to want to know about the performance parameters of the machinery underlying the connections (these would be path properties in the treatment above).

**Comment by theri** [comment]
I agree with expressing pre-establishment and post-establishment properties using the same vocabulary but organizing them differently, i.e., separating them in the API. That's what I was going for in Appendix A.2, perhaps we should refine this further.
So if I understand your suggestion correctly, we would have:

- Connection objects, which have transportParameters representing the application's requirements and preferences, which can only be set pre-establishment.
- protocolStack objects, which have properties expressing what their protocols "can do", i.e., which Protocol Selection Properties they match, what transport features they have. We could name these properties protocolFeatures. They would be expressed in the same language as Protocol Selection Properties, but not have "require"/"prefer"/"avoid"/"prohibit", but just a Boolean "True"/"False" whether they support the given feature. These properties are probably static and the application can only read them to find out what a certain protocolStack is capable of.
- protocolStack objects, which have protocolProperties representing the actual configuration of the protocols, with defaults (e.g., connection timeout for the TCP protocolStack object). At establishment time, these protocolStack objects are bound to the Connection, and the protocolProperties that the application specified pre-establishment are set on these protocolStacks, if applicable. Additionally, protocolProperties can be changed post-establishment.
- path objects, which have pathProperties. I agree that we need this and the application should be allowed to query them at any time. Intuitively, I would think that they are bound to the Connection at establishment time, to reflect the finally chosen path(s). Why bind them to the protocolStack instead?

I'm not yet sure where Intents fit in this, which I guess relates to our discussion in #34.

**Comment by theri** [comment]
I just updated the Appendix A.2 with some text and a table that makes the different types of properties clear: What objects they apply to, when they can be set, and when they can be queried.
I split Protocol Stack into an abstract "Protocol Stack" representing the protocol in general and a "Protocol Stack Instance" representing the protocol as it is actually used in the Connection. I felt like I needed some separation here because it became awkward to always talk about "properties of the protocol in general" vs. "properties that we configure". Feel free to discuss.

In this table, Intents are either per-Connection Transport Parameters, or some of them might also become sendProperties.

"Connection" for the Transport Parameters may become "Connectable" depending on #42. 

Additionally, I removed some default preferences in the table in A.1 because I think it doesn't make sense to have the TAPS system prefer protocols with certain features by default if the application never expressed a wish to have these features. So essentially the defaults for an application that does not specify any Transport Parameters are now just "Require Reliable Data Transfer and In-order delivery".

**Comment by theri** [comment]
I'm going to write a first version of that Querying Section and name it "Maintenance" for now.

**Comment by britram** [comment]
please be brief, we're trying to wrap up work on this to get a -00 out the door early next week, and @mwelzl and I still need to coordinate the review passes...

**Comment by theri** [comment]
Done. Kept this brief, so there is just one getProperties() call for now, but we could also have one for each type of property given there.
If this is compatible with #46, we can close this issue for now, I think.

**Comment by britram** [comment]
LGTM
